systemd/systemd.go

```diff
@@ -202,17 +202,52 @@ type reporter interface {
 }
 
 // New returns a Systemd that uses the given rootDir
-func New(rootDir string, rep reporter) Systemd {
-	return &systemd{rootDir: rootDir, reporter: rep}
-}
+func New(rootDir string, mode InstanceMode, rep reporter) Systemd {
+	return &systemd{rootDir: rootDir, mode: mode, reporter: rep}
+}
+
+// InstanceMode determines which instance of systemd to control.
+//
+// SystemMode refers to the system instance (i.e. pid 1).  UserMode
+// refers to the the instance launched to manage the user's desktop
+// session.  GlobalUserMode controls configuration respected by all
+// user instances on the system.
+//
+// As GlobalUserMode does not refer to a single instance of systemd,
+// some operations are not supported such as starting and stopping
+// daemons.
+type InstanceMode int
+
+const (
+	SystemMode InstanceMode = iota
+	UserMode
+	GlobalUserMode
+)
 
 type systemd struct {
 	rootDir  string
 	reporter reporter
+	mode     InstanceMode
+}
+
+func (s *systemd) systemctl(args ...string) ([]byte, error) {
+	switch s.mode {
+	case SystemMode:
+	case UserMode:
+		args = append([]string{"--user"}, args...)
+	case GlobalUserMode:
+		args = append([]string{"--user", "--global"}, args...)
+	default:
+		panic("unknown InstanceMode")
+	}
+	return systemctlCmd(args...)
 }
 
 // DaemonReload reloads systemd's configuration.
 func (s *systemd) DaemonReload() error {
+	if s.mode == GlobalUserMode {
+		panic("cannot call daemon-reload with GlobalUserMode")
```

Comment:
Should this be an error instead? I'm a bit uneasy about panics. I was thinking about this briefly and was wondering if we could have a collection of types related to systemd, e.g. `systemd.SystemInstance`, `systemd.UserInstance` and `systemd.SettingsForAllUsers` (names are obviously provisional), where each type would only implement the set of operations that make sense and the actual interactions would be through a set of interfaces, exposing them.

I'm not -1 on the PR in the current form, just wanted to voice the idea where a particular error path would not be necessary through the use of the type system.

Reply:
maybe we want different types, I like the idea, but that's a refactor we can do at our leisure.

Reply:
None of the existing uses of this API really have a separation between instantiating the `Systemd` instance and calling its methods, and I'm not sure whether there are many cases where this wouldn't be true.  As this is more of a programming error rather than a runtime error, panics seemed appropriate.

As an alternative to three types, another option would be to split the methods into two types: those that manipulate configuration files vs. those that communicate with a systemd instance.